src/main/scala/elevate/rise/rules/algorithmic.scala

```diff
@@ -107,4 +108,278 @@ object algorithmic {
+  def `x -> join (slide 1 1 x)`: Strategy[Rise] = {
+    x => Success(join(slide(1)(1)(x)) :: x.t)
+  }
+
+  def slideAfter2: Strategy[Rise] = {
+    x => Success(map(fun(x => x `@` lidx(0, 1)), slide(1)(1)(x)) :: x.t)
+  }
+
+  // s -> map snd (zip f s)
+  def zipFstAfter(f: Rise): Strategy[Rise] = s => (f.t, s.t) match {
+    case (ArrayType(n, _), ArrayType(m, _)) if n == m =>
+      Success(map(snd, zip(f, s)) :: s.t)
+    case _ => Failure(zipFstAfter(f))
+  }
+  // f -> map snd (zip f s)
+  def zipSndAfter(s: Rise): Strategy[Rise] = f => (f.t, s.t) match {
+    case (ArrayType(n, _), ArrayType(m, _)) if n == m =>
+      Success(map(fst, zip(f, s)) :: f.t)
+    case _ => Failure(zipFstAfter(s))
+  }
+
+  def dropBeforeJoin: Strategy[Rise] = `J >> drop d -> drop (d / m) >> J >> drop (d % m)`
+  def `J >> drop d -> drop (d / m) >> J >> drop (d % m)`: Strategy[Rise] = {
+    case expr @ App(DepApp(Drop(), d: Nat), App(Join(), in)) => in.t match {
+      case ArrayType(_, ArrayType(m, _)) =>
+        Success(app(drop(d % m), join(drop(d / m)(in))) :: expr.t)
+      case _ => throw new Exception("this should not happen")
+    }
+    case _ => Failure(dropBeforeJoin)
+  }
+
+  // J >> take (n*m - d)
+  // -> dropLast (d / m) >> J >> dropLast (d % m)
+  // -> take (n - d / m) >> J >> take ((n - d / m)*m - d % m)
+  def takeBeforeJoin: Strategy[Rise] = {
+    case expr @ App(DepApp(Take(), nmd: Nat), App(Join(), in)) => in.t match {
+      case ArrayType(n, ArrayType(m, _)) =>
+        val d = n*m - nmd
+        val t1 = n - d / m
+        val t2 = t1*m - d % m
+        Success(app(take(t2), join(take(t1)(in))) :: expr.t)
+      case _ => throw new Exception("this should not happen")
+    }
+    case _ => Failure(dropBeforeJoin)
+  }
+
+  // makeArray(n)(map f1 e)..(map fn e)
+  // -> e |> map(x => makeArray(n)(f1 x)..(fn x)) |> transpose
+  case object mapOutsideMakeArray extends Strategy[Rise] {
+    def matchExpectedMakeArray(mka: Rise): Option[Rise] = mka match {
+      case App(MakeArray(_), App(App(Map(), _), e)) => Some(e)
+      case App(f, App(App(Map(), _), e2)) =>
+        matchExpectedMakeArray(f).flatMap(e =>
+          if (e == e2) { Some(e) } else { None })
+      case _ => None
+    }
+
+    def transformMakeArray(mka: Rise, x: TDSL[Rise]): TDSL[Rise] = mka match {
+      case MakeArray(n) => array(n)
+      case App(mka, App(App(Map(), f), _)) =>
+        app(transformMakeArray(mka, x), app(f, x))
+      case _ => throw new Exception("this should not happen")
+    }
+
+    def apply(expr: Rise): RewriteResult[Rise] =
+      matchExpectedMakeArray(expr) match {
+        case Some(e) => Success(
+          app(transpose, map(fun(x => transformMakeArray(expr, x)))(e))
+            :: expr.t)
+        case None => Failure(mapOutsideMakeArray)
+      }
+  }
+
+  // generate (i => select t (map f e) (map g e))
+  // -> e |> map (x => generate (i => select t (f x) (g x))) |> transpose
+  def mapOutsideGenerateSelect: Strategy[Rise] = {
```

Comment:
this rule is quite specific to arrays of 2 elements built using a combination of generate and select